Refactor apcupsd to use config flow

yuxincs opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
* UI configurations are added to `apcupsd` integration, existing YAML configurations will be automatically imported except the configurations for the `binary_sensor` platform.


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR refactors the `apcupsd` integration to use the config flow framework for better user experience and aligns with home assistant's goals. It also removes YAML config support as directed by architectural decision record (let me know if this is too aggressive and we should first deprecate it). Tested on my APC UPS.

Also, this is my first PR to home assistant, feel free to point out where I did was wrong 😃 

Other code quality changes such adding more tests / type hints will happen in separate PRs after this is merged.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://github.com/home-assistant/core/issues/35230
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24255

With the refactor users can disable polling from HA and setup automations to apply arbitrary polling rate, which is the recommended way to adjust polling rate now IIUC.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

yuxincs committed:
Add Config Flow to APCUPSd integration and remove YAML support.

yuxincs committed:
Hide the binary sensor if user does not select STATFLAG resource.

yuxincs committed:
Add tests for config flows.

yuxincs committed:
Simplify config flow code.

yuxincs committed:
Spell fix.

homeassistant commented:
Hi @yxwangcs, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#64809).

Once you do that we will be able to review and accept this pull request.

Thanks!

emontnemery reviewed (changes requested):
Please add an import flow of the YAML config to not make this a breaking change.

If an import flow is not possible, please explain why.

yuxincs committed:
Fix pylint warnings.

yuxincs committed:
Simplify the code for config flow.

yxwangcs commented:
> Please add an import flow of the YAML config to not make this a breaking change.
> 
> If an import flow is not possible, please explain why.

I was trying to implement import flow for this, the yaml config looks like the following:
```yaml
apcupsd:
  host: localhost
  port: 1234

sensor:
  - platform: apcupsd
     resources:
       - abc
       - efg
       - ....
```

So I have added `async_setup()` to `__init__.py` and from there call 

```python
hass.async_create_task(
        hass.config_entries.flow.async_init(
            DOMAIN, context={"source": SOURCE_IMPORT}, data=config
        )
    )
```

It works fine, however there is a problem when I'm trying to add import flow for the `sensor`.

This PR implements the resource filter (`resources` config) as options, so users can select/deselect which resources they want to integrate into HA. Naturally in order to support import flows I have to add `async_setup_platform()` to `sensor.py` and call 
```python
hass.async_create_task(
            hass.config_entries.options.async_init(
                entry.entry_id,
                context={"source": SOURCE_IMPORT},
                data={CONF_RESOURCES: [res.upper() for res in config[CONF_RESOURCES]]},
            )
        )
```

After doing so, when HA is started for the first time I got the following error:
```
Traceback (most recent call last):
  File "core/homeassistant/config_entries.py", line 327, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/core/homeassistant/components/binary_sensor/__init__.py", line 167, in async_setup_entry
    return await component.async_setup_entry(entry)
  File "core/homeassistant/helpers/entity_component.py", line 157, in async_setup_entry
    raise ValueError("Config entry has already been setup!")
ValueError: Config entry has already been setup!
```
And the resource filter appears in HA but does not seem to function. After restarting HA everything works and looks normal.

Do you know the possible causes for this? Or is there a good guide I can follow for implementing import flows? 😃 

Another point that may be a problem for fully implementing import flows is the `binary_sensor`. In the YAML config you can simply not adding `binary_sensor` section to hide it, but now it's actually controlled by the `STATFLAG` resource option. We cannot simply import this behavior IIUC (`async_setup_platform` is not even called).

yuxincs committed:
First attempt to implement import flows to suppport legacy YAML configurations.

yxwangcs commented:
BTW is there a guide on implementing import flows to set options? I tried to call `hass.config_entries.options.async_init` (or `hass.config_entries.async_update_entry`) in sensor's `async_setup_platform` to update the `resources` options, but it's giving me weird bugs that seem like race conditions:

```
  File "/core/homeassistant/config_entries.py", line 466, in async_unload
    result = await component.async_unload_entry(hass, self)
  File "/core/homeassistant/components/binary_sensor/__init__.py", line 173, in async_unload_entry
    return await component.async_unload_entry(entry)
  File "/core/homeassistant/helpers/entity_component.py", line 172, in async_unload_entry
    raise ValueError("Config entry was never loaded!")
ValueError: Config entry was never loaded!
```

```
 File "/core/homeassistant/config_entries.py", line 327, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/core/homeassistant/components/sensor/__init__.py", line 210, in async_setup_entry
    return await component.async_setup_entry(entry)
  File "/core/homeassistant/helpers/entity_component.py", line 157, in async_setup_entry
    raise ValueError("Config entry has already been setup!")
ValueError: Config entry has already been setup!
```

I added an artificial sleep before `hass.config_entries.options.async_init` and the errors disappear, I'm guessing I have to wait for platforms to be fully set up before I can update the options in the config entry (since it will trigger a config reload, which in turn will trigger a re-setup)? But I'm not completely sure... [this](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwjjwo7J48v1AhXEc98KHW0BBeoQFnoECAUQAQ&url=https%3A%2F%2Fcommunity.home-assistant.io%2Ft%2Fconfig-flow-import-set-options%2F198492&usg=AOvVaw2nQKToh7ZeeGO6W2ZyTop4) is the most related article I can find but it does not mention bugs like this....

yuxincs committed:
Remove unnecessary log calls.

yuxincs committed:
Wrap synchronous update call with `hass.async_add_executor_job`.

yuxincs committed:
Import the YAML configurations when sensor platform is set up.

yuxincs committed:
Move the logger call since the variables are not properly set up.

yxwangcs commented:
OK I think I have found a workaround for the issue above. We should actually set up the config entry (`hass.config_entries.flow.async_init`) when we have all the information (`host`, `port` and `resources`). 

So when we set up the integration we temporarily store the `host` and `port` information in `hass.data[DOMAIN]` with a special entry ID `SOURCE_IMPORT` (to avoid conflicts with existing entry IDs). Later when sensor platform is set up, we gather the `resources` field and from there we import the config entry. The config flow is also updated to import extra `CONF_RESOURCES` field in `user_input` as options as well (though it does not show in the form). 

Tested locally and it works great. This PR is now ready for review with a caveat: the YAML config for `binary_sensor` is not imported (should we log a warning?) because of the workaround, also it only has a `name` field which can be easily configured through UI I think....

yuxincs committed:
Add codeowner.

sagitt commented:
> OK I think I have found a workaround for the above issue. We should actually set up the config entry (`hass.config_entries.flow.async_init`) when we have all the information (`host`, `port` and `resources`). 
> 
> 
> 
> So when we set up the integration we temporarily store the `host` and `port` information in `hass.data[DOMAIN]` with a special entry ID `SOURCE_IMPORT` (to avoid conflicts with existing entry IDs). Later when sensor platform is set up, we gather the `resources` field and from there we import the config entry. The config flow is also updated to import extra `CONF_RESOURCES` field in `user_input` as options as well (though it does not show in the form). It's indeed a bit hacky but once we fully deprecate YAML config in the near future it's not hard to remove this.
> 
> 
> 
> Tested locally and it works great. This PR is now ready for review with a caveat: the YAML config for `binary_sensor` is not imported (should we log a warning?) because of the workaround, but it only has a `name` field which can be easily configured through UI I think....

Thank you for your hard work. From next release apcupsd will be disabled. I hope soon this update and a working version. :)

MartinHjelmare commented:
Yes, we've disabled the integration as its requirement library isn't compatible with Python 3.10.

https://github.com/home-assistant/core/blob/406fbca4bce3c3be9d84a4b59eb40f4a9726a3ff/homeassistant/components/apcupsd/manifest.json#L2

This will need to be addressed and solved before this PR can continue review and possible merge.

sagitt commented:
> Yes, we've disabled the integration as its requirement library isn't compatible with Python 3.10.
> 
> 
> 
> https://github.com/home-assistant/core/blob/406fbca4bce3c3be9d84a4b59eb40f4a9726a3ff/homeassistant/components/apcupsd/manifest.json#L2
> 
> 
> 
> This will need to be addressed and solved before this PR can continue review and possible merge.

This is a BIG problem, i (and lot others people) use apcupsd and without it lots of automations break (security automation too).

Maybe there is a workaround?

yxwangcs commented:
> > Yes, we've disabled the integration as its requirement library isn't compatible with Python 3.10.
> > https://github.com/home-assistant/core/blob/406fbca4bce3c3be9d84a4b59eb40f4a9726a3ff/homeassistant/components/apcupsd/manifest.json#L2
> > 
> > This will need to be addressed and solved before this PR can continue review and possible merge.
> 
> This is a BIG problem, i (and lot others people) use apcupsd and without it lots of automations break (security automation too).
> 
> Maybe there is a workaround?

I haven't looked at `apcaccess` library to see where the incompatibilities come from, but I guess the only workaround is just not upgrade until this is resolved.

I'll find some time to create a PR to `apcaccess` to fix this, hopefully soon ;)

pailloM commented:
> Can someone else verify the incompatibilities between python 3.10 and `apcaccess`? I have rebased this branch onto the current dev, but seems like all tests run fine. I ran a local instance and everything seems to work....
> 
> I also ran the tests for `apcaccess` library as well with python 3.10 and there are no errors or warnings. There are no explanations in the PR (#59729) that disabled it, so it would be nice if there is more information.
> 
> But anyways, perhaps a newer and async version of `apcaccess` should be implemented for performance, and we can drop `apcaccess`.

I have asked this same question on the beta release discord channel. @frenck said he would check.
apcaccess installs fine in python 3.10, is a pure python library and has no extra requirements. I moved the integration to a customs components and it loads fine with the beta 1 release. I don't really understand where the incompatibility is.
For sure it has not been updated in a long time and could do with some async love.

pailloM commented:
The integration has been re-enabled in 2022.3.0b4.
In this PR the name field in the manifest.json is not correct and the requirements_test_all.txt adds dependencies for another integration. You might not have the latest version of these 2 files.
Thanks for your great work.

yuxincs committed:
Fix name field of manifest.json.

yuxincs committed:
Fix linting issue.

yuxincs committed:
Fix incorrect dependency due to incorrect rebase.

yxwangcs commented:
> The integration has been re-enabled in 2022.3.0b4. In this PR the name field in the manifest.json is not correct and the requirements_test_all.txt adds dependencies for another integration. You might not have the latest version of these 2 files. Thanks for your great work.

Just fixed and rebased onto the latest `dev` branch, this PR is now ready for review, thanks! :)

calas80 commented:
> > Yes, we've disabled the integration as its requirement library isn't compatible with Python 3.10.
> > https://github.com/home-assistant/core/blob/406fbca4bce3c3be9d84a4b59eb40f4a9726a3ff/homeassistant/components/apcupsd/manifest.json#L2
> > 
> > This will need to be addressed and solved before this PR can continue review and possible merge.
> 
> This is a BIG problem, i (and lot others people) use apcupsd and without it lots of automations break (security automation too).
> 
> Maybe there is a workaround?

Also for me this is a huge problem! is there any news?

yxwangcs commented:
> > > Yes, we've disabled the integration as its requirement library isn't compatible with Python 3.10.
> > > https://github.com/home-assistant/core/blob/406fbca4bce3c3be9d84a4b59eb40f4a9726a3ff/homeassistant/components/apcupsd/manifest.json#L2
> > > 
> > > This will need to be addressed and solved before this PR can continue review and possible merge.
> > 
> > 
> > This is a BIG problem, i (and lot others people) use apcupsd and without it lots of automations break (security automation too).
> > Maybe there is a workaround?
> 
> Also for me this is a huge problem! is there any news?

I think 
> The integration has been re-enabled in 2022.3.0b4.

So this won't be a problem, at least for now....

yuxincs committed:
Update codeowner and config flows via hassfest.

yuxincs committed:
Postpone the deprecation warning to 2022.7.

yxwangcs commented:
Rebased onto the current `dev` branch and postponed the deprecation notice from 2022.4 to 2022.7.

yuxincs committed:
Import future annotations for init file.

yuxincs committed:
Add an newline at the end to make prettier happy.

yuxincs committed:
Update github id.

homeassistant commented:
Hi @yuxincs, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#64809).

Once you do that we will be able to review and accept this pull request.

Thanks!

yuxincs reviewed (commented):
Import flow is properly implemented and this PR is rebased onto the current `dev` branch, ready to be reviewed 😃

yuxincs committed:
Add type hints for return types of steps in config flow.

yuxincs committed:
Move the deprecation date for YAML config to 2022.12.

MartinHjelmare commented on homeassistant/components/apcupsd/__init__.py:
```diff
@@ -1,60 +1,130 @@
+    if (conf := config.get(DOMAIN)) is None:
+        return True
+
+    _LOGGER.warning(
```
We only need one warning when importing config YAML. Please consolidate this warning with the warning in the sensor platform.

yuxincs replied:
Done!

MartinHjelmare commented on homeassistant/components/apcupsd/__init__.py:
```diff
@@ -1,60 +1,130 @@
+    hass.data[DOMAIN][config_entry.entry_id] = data_service
+
+    # Forward the config entries to the supported platforms.
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, platform)
+        )
```
```suggestion
    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
```

MartinHjelmare commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+                status: dict[str, Any] | None = await self.hass.async_add_executor_job(
+                    lambda: data_service.status
+                )
+                if status is None:
```
Please only wrap the line that can raise inside the try... except block. Use an `else:` block if needed.

yuxincs replied:
Done (this part is rewritten)

MartinHjelmare commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+                # them out and log warnings for them.
+                options = None
+                if self.source == SOURCE_IMPORT and CONF_RESOURCES in user_input:
+                    available_resources = set(data_service.status.keys())
```
```suggestion
                    available_resources = set(data_service.status)
```

MartinHjelmare commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+                            continue
+                        resources.append(resource)
+
+                    options = {CONF_RESOURCES: resources}
```
I'd update the item in the config entry data instead, since we won't be adding an options flow.

yuxincs replied:
Yes, merged this into config entry data.

MartinHjelmare commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+        schema = vol.Schema(
+            {
+                vol.Required(CONF_HOST, default="localhost"): cv.string,
+                vol.Required(CONF_PORT, default=3551): cv.port,
```
Use a selector instead for the port.

Example:
https://github.com/home-assistant/core/blob/cd1227d8b9c4ae677f1218aeabc9dea497393db9/homeassistant/components/mysensors/config_flow.py#L48-L55

yuxincs replied:
Done!

MartinHjelmare commented on homeassistant/components/apcupsd/manifest.json:
```diff
@@ -1,9 +1,11 @@
   "name": "apcupsd",
+  "config_flow": true,
   "documentation": "https://www.home-assistant.io/integrations/apcupsd",
+  "dependencies": [],
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/apcupsd/__init__.py:
```diff
@@ -1,60 +1,130 @@
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
 class APCUPSdData:
```
Please rewrite `APCUPSdData` so it doesn't call a library method in the `status` property. We shouldn't do I/O when accessing properties. API calls should be explicit.

yuxincs replied:
Since this PR was designed to add config flow, the original code was intentionally left untouched. But since we are doing a lot already, `APCUPSdData` class is simplified and now any I/O is explicit.

yuxincs committed:
Update according to reviews.

yuxincs committed:
Use async_forward_entry_setups.

yuxincs committed:
Add helper properties to `APCUPSdData` class.

yuxincs committed:
Add device_info for binary sensor.

yuxincs committed:
Simplify config flow.

yuxincs committed:
Remove options flow strings.

yuxincs committed:
update the tests according to the changes.

yuxincs committed:
Add `entity_registry_enabled_default` to entities and use imported CONF_RESOURCES to disable entities instead of skipping them.

yuxincs commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+
+        schema = vol.Schema(
+            {
+                vol.Optional(CONF_RESOURCES, default=selected): cv.multi_select(
```
Updated to do this instead of options flow 👍

MartinHjelmare commented on homeassistant/components/apcupsd/__init__.py:
```diff
@@ -1,60 +1,103 @@
-    except Exception:  # pylint: disable=broad-except
-        _LOGGER.exception("Failure while testing APCUPSd status retrieval")
+        await hass.async_add_executor_job(data_service.update)
+    # It doesn't really matter why we're not able to get the status, just that we can't.
```
This is not a good comment. We try to avoid catching broad exceptions after the config flow in integrations. Preferably this except handler should be narrowed down in the future.

Eg, if we fail to connect we may want to try again later. If there's a user error reason why it failed we need to ask the user to resolve that and not retry.

yuxincs replied:
I agree! For sake of simplicity of this PR (it is already pretty complex). I can create a separate PR to properly fix this when this PR is merged (along with a PR to add more thorough tests of the entire integration). 

Or if you prefer, I can try to fix this in this PR as well 😃

MartinHjelmare replied:
Let's remove the comment.
```suggestion
```

MartinHjelmare commented on homeassistant/components/apcupsd/binary_sensor.py:
```diff
@@ -1,43 +1,82 @@
-    apcups_data = hass.data[DOMAIN]
+    data_service: APCUPSdData = hass.data[DOMAIN][config_entry.entry_id]
+
+    description = BinarySensorEntityDescription(
```
If the description is static we should move it to a constant at the module level.

yuxincs replied:
Agreed, fixed!

MartinHjelmare commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+                step_id="user", data_schema=schema, errors=errors
+            )
+
+        if len(data_service.status) == 0:
```
```suggestion
        if not data_service.status:
```

MartinHjelmare commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+            return self.async_abort(reason="no_status")
+
+        # We _try_ to use the serial number of the UPS as the unique id since
+        # this field is not guaranteed to exist on all APC UPS models.
```
Maybe also guard against more than one config entry with the same host and port, after checking for unique_id abort?

https://github.com/home-assistant/core/blob/cf5a11a1e7c7ef49a439753f3bb708d10b2c22b9/homeassistant/config_entries.py#L1281-L1285

yuxincs replied:
Wasn't aware of this, added the guard! 

It is impossible to have two apcupsd integrations with same host and port (while having different `unique_id`s), so I feel it is better to put the guard right after `user_input` is not `None`. 

(If a system has multiple APC UPSes, [there must be multiple instances of apcupsd listening on different addresses](https://wiki.debian.org/apcupsd#Set_up_the_apcupsd_configuration_files))

Let me know if you think putting it after unique_id abort is still a good idea.

MartinHjelmare commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+    """APCUPSd integration config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_LOCAL_POLL
```
This has moved to the manifest.json.
```suggestion
```

MartinHjelmare commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+
+from . import DOMAIN, APCUPSdData
+
+_LOGGER = logging.getLogger(__name__)
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/apcupsd/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+"""Config flow for APCUPSd integration."""
+from __future__ import annotations
+
+import logging
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/apcupsd/strings.json:
```diff
@@ -0,0 +1,20 @@
+          "host": "[%key:common::config_flow::data::host%]",
+          "port": "[%key:common::config_flow::data::port%]"
+        },
+        "description": "Enter the [%key:common::config_flow::data::host%] and [%key:common::config_flow::data::port%] on which the apcupsd NIS is being served."
```
We don't mix in references inside sentences as that may not work well for all languages.

MartinHjelmare commented on tests/components/apcupsd/test_config_flow.py:
```diff
@@ -0,0 +1,168 @@
+        for status in (MOCK_STATUS, MOCK_MINIMAL_STATUS):
+            mock_parse.return_value = status
+
+            result = await hass.config_entries.flow.async_init(
```
Please patch the integration setup function on all create entry results to avoid setting up the whole integration.

See the scaffold script test template for example.

yuxincs replied:
I see, added the patch. 

Actually in the latest revision, I partially rewrote the tests for the config flow to have 100% coverage with simpler structure. Hope that also makes sense to you.

MartinHjelmare commented on tests/components/apcupsd/test_config_flow.py:
```diff
@@ -0,0 +1,168 @@
+    )
+
+
+async def test_flow_works(hass: HomeAssistant, config_entry: MockConfigEntry):
```
Please type the whole signature when adding type annotations.
```suggestion
async def test_flow_works(hass: HomeAssistant, config_entry: MockConfigEntry) -> None:
```

yuxincs replied:
Got it, made changes to all test functions :)

MartinHjelmare commented on homeassistant/components/apcupsd/sensor.py:
```diff
@@ -406,36 +429,92 @@
+    # options (although not shown in UI during config setup).
+    conf[CONF_RESOURCES] = config[CONF_RESOURCES]
+
+    _LOGGER.warning(
```
```suggestion
    _LOGGER.debug(
```

MartinHjelmare commented on homeassistant/components/apcupsd/sensor.py:
```diff
@@ -406,36 +429,92 @@
+            _LOGGER.warning("Invalid resource from APCUPSd: %s", resource.upper())
+            continue
+
+        description = copy.copy(SENSORS[resource])
```
Don't copy the description. Pass an extra parameter to represent enabled entry to the entity for the imported config data instead. Setting `_attr_entity_registry_enabled_default` takes precedence over the entity description setting.

yuxincs committed:
Update according to reviews.

yuxincs committed:
Do not use model of the UPS as the title for the integration.

Instead, simply use "APCUPSd" as the integration title and let the device info serve as title for each device instead.

yuxincs committed:
Change schema to be a global variable.

yuxincs committed:
Add more comments.

yuxincs committed:
Rewrite the tests for config flows.

yuxincs committed:
Fix enabled_by_default.

yuxincs committed:
Show friendly titles in the integration.

yuxincs committed:
Add import check in `async_setup_platform` to avoid importing in sensor platform setup.

yuxincs committed:
Add import check in `async_setup_platform` to avoid importing in sensor platform setup.

yuxincs committed:
Update comments in test files.

MartinHjelmare commented on homeassistant/components/apcupsd/__init__.py:
```diff
@@ -1,87 +1,156 @@
-MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+DOMAIN: Final = "apcupsd"
+VALUE_ONLINE: Final = 8
+PLATFORMS: Final = (Platform.SENSOR, Platform.BINARY_SENSOR)
```
```suggestion
PLATFORMS: Final = (Platform.BINARY_SENSOR, Platform.SENSOR)
```

MartinHjelmare commented on homeassistant/components/apcupsd/binary_sensor.py:
```diff
@@ -1,43 +1,81 @@
+            return None
+
+        return {
+            "identifiers": {(DOMAIN, self._data_service.model)},
```
The model is not a unique id for the device. We should probably use the config entry entry_id.

yuxincs replied:
We can actually use `serial_no` if that is available. I think that better suits the purpose here?

MartinHjelmare commented on homeassistant/components/apcupsd/binary_sensor.py:
```diff
@@ -1,43 +1,81 @@
+        self._data_service = data_service
+
+    @property
+    def device_info(self) -> DeviceInfo | None:
```
We can set `self._attr_device_info` in the init method instead.

yuxincs replied:
Updated!

MartinHjelmare commented on homeassistant/components/apcupsd/sensor.py:
```diff
@@ -454,18 +539,44 @@ def infer_unit(value):
+        self._data_service = data_service
+
+    @property
+    def device_info(self) -> DeviceInfo | None:
```
As above.

yuxincs replied:
Updated!

MartinHjelmare commented on tests/components/apcupsd/test_config_flow.py:
```diff
@@ -0,0 +1,216 @@
+    ), _patch_setup() as mock_setup:
+        # We test different combinations of minimal statuses, where the title of the
+        # integration will vary.
+        extras = (
```
I'd move these items to a parameter and use `pytest.mark.parametrize`.

https://docs.pytest.org/en/stable/how-to/parametrize.html

yuxincs replied:
Agreed! This made this test much cleaner. 👍

MartinHjelmare commented on tests/components/apcupsd/test_config_flow.py:
```diff
@@ -0,0 +1,216 @@
+        mock_setup.assert_called_once()
+
+
+async def test_flow_import_with_invalid_resource(hass: HomeAssistant) -> None:
```
I don't think we check any resources in the config flow.

yuxincs replied:
Right, this test is actually testing the integration instead of config flow. 

Removed, and will add this back in a separate PR.

MartinHjelmare commented on homeassistant/components/apcupsd/__init__.py:
```diff
@@ -1,87 +1,156 @@
-        _LOGGER.exception("Failure while testing APCUPSd status retrieval")
+        await hass.async_add_executor_job(data_service.update)
+    # It doesn't really matter why we're not able to get the status, just that we can't.
+    except Exception as ex:  # pylint: disable=broad-except
```
We only catch OSError in the config flow.
```suggestion
    except OSError as ex:
```

MartinHjelmare commented on homeassistant/components/apcupsd/__init__.py:
```diff
@@ -1,87 +1,156 @@
+        await hass.async_add_executor_job(data_service.update)
+    # It doesn't really matter why we're not able to get the status, just that we can't.
+    except Exception as ex:  # pylint: disable=broad-except
+        _LOGGER.exception("Failure while testing APCUPSd status retrieval", exc_info=ex)
```
```suggestion
        _LOGGER.error("Failure while testing APCUPSd status retrieval: %s", ex)
```

yuxincs committed:
Use parametrize instead of manually iterating different test cases.

yuxincs committed:
Swap the order of the platform constants.

yuxincs committed:
Avoid using broad exceptions.

yuxincs committed:
Set up device info via `_attr_device_info`.

yuxincs committed:
Remove unrelated test in `test_config_flow`.

MartinHjelmare commented on homeassistant/components/apcupsd/binary_sensor.py:
```diff
@@ -1,43 +1,74 @@
+        # Set up unique id and device info if serial number is available.
+        if (serial_no := data_service.serial_no) is not None:
+            self._attr_unique_id = f"{serial_no}_{description.key}"
+            self._attr_device_info = {
+                "identifiers": {(DOMAIN, serial_no)},
+                "model": data_service.model,
+                "manufacturer": "APC",
+                "hw_version": data_service.hw_version,
+                "sw_version": data_service.sw_version,
+            }
```
```suggestion
            self._attr_device_info = DeviceInfo(
                identifiers={(DOMAIN, serial_no)},
                model=data_service.model,
                manufacturer="APC",
                hw_version=data_service.hw_version,
                sw_version=data_service.sw_version,
            )
```

MartinHjelmare commented on homeassistant/components/apcupsd/binary_sensor.py:
```diff
@@ -1,43 +1,74 @@
+from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant
-import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
```
```suggestion
from homeassistant.helpers.entity import DeviceInfo
from homeassistant.helpers.entity_platform import AddEntitiesCallback
```

MartinHjelmare commented on homeassistant/components/apcupsd/sensor.py:
```diff
@@ -29,365 +32,383 @@
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
```
```suggestion
from homeassistant.helpers.entity import DeviceInfo
from homeassistant.helpers.entity_platform import AddEntitiesCallback
```

MartinHjelmare commented on homeassistant/components/apcupsd/sensor.py:
```diff
@@ -454,18 +538,38 @@ def infer_unit(value):
+        # Set up unique id and device info if serial number is available.
+        if (serial_no := data_service.serial_no) is not None:
+            self._attr_unique_id = f"{serial_no}_{description.key}"
+            self._attr_device_info = {
+                "model": data_service.model,
+                "manufacturer": "APC",
+                "hw_version": data_service.hw_version,
+                "sw_version": data_service.sw_version,
+            }
```
```suggestion
            self._attr_device_info = DeviceInfo(
                model=data_service.model,
                manufacturer="APC",
                hw_version=data_service.hw_version,
                sw_version=data_service.sw_version,
            )
```

MartinHjelmare replied:
I recommend setting a name here and following our entity naming guidelines:

https://developers.home-assistant.io/docs/core/entity#entity-naming

yuxincs replied:
We have already set the `entity_description` in this entity, which contains a `name` field. From reading https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/entity.py#L323-L324, it seems that it will fall back to using the `entity_description.name` if a name is not explicitly set.

I'm a bit confused here as to which is recommended, actually.  I can get rid of `entity_description` if it is discouraged (since I didn't see it mentioned in https://developers.home-assistant.io/docs/core/entity)

Also, all the names in current entity description objects contain a `UPS` prefix, I assume this is also discouraged from reading the doc you referenced, but I didn't change them due to compatibility. Do you think we should bring a breaking change for this?

MartinHjelmare replied:
Since there's no unique_id set for the entities before this PR we should not change names to the new style now but do that after two releases when we remove the import of the YAML config.

yuxincs committed:
Use `DeviceInfo` instead of dict to assign to `_attr_device_info`.

yuxincs committed:
Add english translation.

MartinHjelmare commented on homeassistant/components/apcupsd/sensor.py:
```diff
@@ -406,36 +428,99 @@
+    if conf is None:
+        return
+
+    _LOGGER.warning(
```
Please also create a repairs issue.

Example:
https://github.com/home-assistant/core/blob/499c3410d1177eeec478af366e275a41b3e6ea60/homeassistant/components/openexchangerates/sensor.py#L42-L50

https://github.com/home-assistant/core/blob/499c3410d1177eeec478af366e275a41b3e6ea60/homeassistant/components/openexchangerates/strings.json#L27-L32

yuxincs replied:
Done 🎉 !

MartinHjelmare reviewed (commented):
One final comment. Then we're good.

yuxincs committed:
Add `async_create_issue` for deprecated YAML configuration.

yuxincs committed:
Enable UPS status by default since it could show "online, charging, on battery etc" which is meaningful for all users.

MartinHjelmare commented on homeassistant/components/apcupsd/sensor.py:
```diff
@@ -461,6 +462,16 @@ async def async_setup_platform(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        breaks_in_ha_version="2022.10.0",
```
Our normal deprecation period is two releases.
```suggestion
        breaks_in_ha_version="2022.12.0",
```

MartinHjelmare commented on homeassistant/components/apcupsd/strings.json:
```diff
@@ -16,5 +16,11 @@
+  },
+  "issues": {
+    "deprecated_yaml": {
+      "title": "The apcupsd YAML configuration is being removed",
```
```suggestion
      "title": "The APC UPS Daemon YAML configuration is being removed",
```

MartinHjelmare commented on homeassistant/components/apcupsd/strings.json:
```diff
@@ -16,5 +16,11 @@
+  "issues": {
+    "deprecated_yaml": {
+      "title": "The apcupsd YAML configuration is being removed",
+      "description": "Configuring apcupsd using YAML is being removed.\n\nYour existing YAML configuration has been imported into the UI automatically.\n\nRemove the apcupsd YAML configuration from your configuration.yaml file and restart Home Assistant to fix this issue."
```
```suggestion
      "description": "Configuring APC UPS Daemon using YAML is being removed.\n\nYour existing YAML configuration has been imported into the UI automatically.\n\nRemove the APC UPS Daemon YAML configuration from your configuration.yaml file and restart Home Assistant to fix this issue."
```

MartinHjelmare commented on homeassistant/components/apcupsd/translations/en.json:
```diff
@@ -0,0 +1,26 @@
+    },
+    "issues": {
+        "deprecated_yaml": {
+            "description": "Configuring apcupsd using YAML is being removed.\n\nYour existing YAML configuration has been imported into the UI automatically.\n\nRemove the apcupsd YAML configuration from your configuration.yaml file and restart Home Assistant to fix this issue.",
```
```suggestion
            "description": "Configuring APC UPS Daemon using YAML is being removed.\n\nYour existing YAML configuration has been imported into the UI automatically.\n\nRemove the APC UPS Daemon YAML configuration from your configuration.yaml file and restart Home Assistant to fix this issue.",
```

MartinHjelmare committed:
Apply suggestions from code review

MartinHjelmare committed:
Apply suggestion

MartinHjelmare committed:
Apply suggestion
